File: src/two_sum/TwoSum.java
package two_sum;

import java.util.HashMap;
import java.util.Map;

/**
 * two sum https://leetcode.com/problems/two-sum/description/
 * Created by gavingeng on 2018-05-23 16:37.
 */
public class TwoSum {

	/**
	 * 时间复杂度 O(n^2)
	 * 空间复杂度 O(1)
	 */
	public static int[] solution1(int[] nums, int target) {
		int[] idxResult = {0, 0};
		for (int i = 0; i < nums.length - 1; i++) {
			for (int j = i + 1; j < nums.length; j++) {
				if (nums[i] + nums[j] == target) {
					idxResult[0] = i;
					idxResult[1] = j;
				}
			}
		}
		return idxResult;
	}

	public static int[] solution2(int[] nums, int target) {
		int[] idxResult = {0, 0};
		//key:值 value:索引
		Map<Integer, Integer> map = new HashMap<>();
		for (int i = 0; i < nums.length; i++) {
			map.put(nums[i], i);
		}
		for (int i = 0; i < nums.length; i++) {
			int aim = target - nums[i];
			if (map.containsKey(aim)) {
				idxResult[0] = i;
				idxResult[1] = map.get(aim);
			}
		}
		return idxResult;
	}


	public static void main(String[] args) {
		int[] nums = {2, 7, 11, 15};
		int target = 9;

//		int[] idxResult = solution1(nums, target);
		int[] idxResult = solution2(nums, target);

		for (int i = 0; i < idxResult.length; i++) {
			System.out.println(idxResult[i]);
		}
	}
}

File: src/two_sum/twoSum.go
package main

import "fmt"

func solution1(nums []int, target int) []int {
	ret := []int{0, 0}
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if target-nums[j] == nums[i] {
				ret[0] = i
				ret[1] = j
				return ret
			}
		}
	}

	return ret
}

func solution2(nums []int, target int) []int {
	ret := []int{0, 0}
	data := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		data[nums[i]] = i
	}

	for i := 0; i < len(nums); i++ {
		tar := target - nums[i]
		v, exists := data[tar]
		if exists {
			ret[0] = i
			ret[1] = v
			return ret
		}
	}
	return ret

}

func solution3(nums []int, target int) []int {
	ret := []int{0, 0}
	data := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		tar := target - nums[i]
		v, exist := data[tar]
		if exist {
			ret[0] = i
			ret[1] = v
			return ret
		} else {
			data[nums[i]] = i
		}
	}

	return ret
}

func printArray(ret []int) {
	for i := 0; i < len(ret); i++ {
		println("i=", i, "|data=", ret[i])
	}
}

func main() {
	fmt.Println("two-sum")
	nums := []int{2, 7, 11, 15}
	target := 9

	//ret := solution1(nums, target)
	//ret := solution2(nums, target)
	ret := solution3(nums, target)

	printArray(ret)
}
